public/utils.ts

```diff
@@ -0,0 +1,262 @@
+  if (lastPart !== undefined) {
+    delete current[lastPart];
+  }
+
+  return obj;
+};
+
+export const getFromDict = (obj: object, key: string): any => {
+  let keyParts: string[] = key.split(".");
+  let current: any = obj;
+
+  for (let part of keyParts) {
+    if (current instanceof Map) {
+      if (!current.has(part)) {
+        return null;
+      }
+      current = current.get(part);
+    } else if (typeof current === "object" && current !== null) {
+      if (!(part in current)) {
+        return null;
+      }
+      current = current[part];
+    } else {
+      return null;
+    }
+  }
+
+  return current;
+};
+
+export const setToDict = (
+  container: { [key: string]: any } | Map<any, any>,
+  key: string,
+  value: any,
+) => {
+  const keys: string[] = key.trim().split(".");
+  let current_container = container;
+
+  for (let [ind, currentKey] of keys.entries()) {
+    if (ind === keys.length - 1) {
+      if (current_container instanceof Map) {
+        current_container.set(currentKey, value);
+      } else {
+        (current_container as { [key: string]: any })[currentKey] = value;
+      }
+    } else {
+      if (current_container instanceof Map) {
+        if (
+          !current_container.has(currentKey) ||
+          !(current_container.get(currentKey) instanceof Map)
+        ) {
+          current_container.set(currentKey, new Map<any, any>());
+        }
+        current_container = current_container.get(currentKey);
+      } else {
+        if (
+          !current_container[currentKey] ||
+          typeof current_container[currentKey] !== "object"
+        ) {
+          (current_container as { [key: string]: any })[currentKey] = {};
+        }
+        current_container = current_container[currentKey];
+      }
+    }
+  }
+
+  return container;
+};
+
+export const generateRangeFromStep = (
+  start: number,
+  end: number,
+  step: number,
+): number[] => {
+  if (step <= 0) {
+    throw new Error("Step must be greater than 0");
+  }
+  if (start > end && step > 0) {
+    throw new Error("Start cannot be greater than end when step is positive");
+  }
+
+  let returnArray: number[] = [];
+  let track = start;
+  while (track <= end) {
+    returnArray.push(track);
+    track = roundTwoDecimals(track + step);
+  }
+
+  return returnArray;
+};
+
+export const generateRangeFromSlots = (
+  start: number,
+  end: number,
+  slots: number,
+): number[] => {
+  let returnArray: number[] = [];
+
+  if (slots === 0) {
+    return returnArray;
+  }
+  if (slots === 1) {
+    return [start, end];
+  }
+  if (start === end) {
+    return [start];
+  }
+  if (start > end) {
+    [start, end] = [end, start];
+  }
+
+  let step = (end - start) / slots;
+
+  for (let i = start; i <= end; i += step) {
+    returnArray.push(roundTwoDecimals(i));
+  }
+
+  return returnArray;
+};
+
+export const distinct = (arr: any[]): any[] => {
+  // @ts-ignore
+  return [...new Set<any>(arr)];
+};
```

Comment:
Better for TypeScript:

```
export const distinct = <T>(arr: T[]): T[] => {
  return [...new Set(arr)];
};
```

Then the return type is the same as the input type, rather than any[]